doc/2/plugins/guides/hooks/index.md

```diff
@@ -15,6 +15,116 @@ Hooks can only listen: the received information cannot be changed. And Kuzzle do
 
 ## Usage
 
+<SinceBadge version="2.2.0" />
+
+Plugins can register hooks by exposing a `hooks` object: keys are listened [events](/core/2/plugins/guides/events), and values are configuration object to describe the hook behavior.
+
+Hook configuration objects contains 2 properties:
+ - `callback`: a function that will be called with a payload related to the event
+ - `filter`: an optional Koncorde filter conditioning the execution of the callback
+
+Hook declaration structure:
+```js
+this.hooks = {
+  'event-name': {
+    callback: <fn>,
+    filter: { ... }
+  }
+};
+```
+
+### Unconditional listening
+
+If no filter is set, then the callback is called each time the event is triggered.  
+The callback arguments depend on the selected event. 
+
+```js
+this.hooks = {
+  'collection:afterList': {
+    callback: request => {
+      // this callback will be called each time the event is triggered
+    }
+  },
+  'document:generic:beforeWrite': {
+    callback: (documents, request) => {
+      // this callback will be called each time the event is triggered
+    }
+  }
+};
+```
+
+### Conditional listening
+
+If a filter is defined, then its application depends on the type of the first argument usually passed to the callback.  
+
+::: warning
+Filters are only compatible with events that have an object or an array of objects as the first argument.
+:::
+
+#### Object as first event arguments
+
+If the first argument of the event is an object, then the filter is applied to that object.
+
+When the object matches the chosen filter, then the callback is called with the usual arguments of the event.
+
+Example with an [API event](/core/2/plugins/guides/events/api-events/):
+
+```js
+this.hooks = {
+  'collection:afterList': {
+    callback: request => {
+      // this callback will be called only 
+      // when request.input.resource.index is equals to 'nyc-open-data'
+    },
+    filter: {
+      equals: { 'input.resource.index': 'nyc-open-data' }
+    }
+  }
+}
+```
+
+#### Object array as first event argument
+
+If the first argument of the event is an array of objects, then the filter is applied on each object.  
+
+When an object matches the chosen filter, then the callback is called with that object as the first parameter, and if the event had other arguments then they will also  be passed.
+
+Example with a [Document generic event](/core/2/plugins/guides/events/generic-document-events/):
+```js
+this.hooks = {
+  'document:generic:beforeWrite': {
+    callback: (document, request) => {
+      // this callback will be called only when a document contains
+      // a 'name' property equals to 'Saigon'
+    },
+    filter: {
+      equals: { '_source.city': 'Saigon' }
+    }
+  }
+}
+```
```

Comment:
Wouldn't it be better to invoke the callback with a filtered array containing only objects matching the filters?

```suggestion
The callback is invoked if at least one object of the array matches the filters. The callback payload will be an array containing only matching objects.

Example with a [Document generic event](/core/2/plugins/guides/events/generic-document-events/):
```js
this.hooks = {
  'document:generic:beforeWrite': {
    callback: (document, request) => {
      // this callback will be called only when at least one document contains
      // a 'name' property equal to 'Saigon'
      // "document" will be an array of objects containing only a "name" property
      // equal to "Saigon"
    },
    filter: {
      equals: { '_source.city': 'Saigon' }
    }
  }
}
```
```

Reply:
... I think I know why you worded it like you did. Because pipes/hooks can be chained, with payloads being passed from a plugin to another. :expressionless:

Reply:
Yes for the hooks it could be a filtered array but for pipes you have to return an array which will be passed to the next pipe :disappointed:

Reply:
This can still be done I think, and it would make this feature really nice.

Imagine our pipe runner having a pipe wrapper for object arrays, behaving like this (rough approximation, I didn't study our current code in details):

```js
function wrapObjectArrays (arr, cb, next) {
  const filtered = [];
  const indices = [];

  for (let i = 0; i < arr.length; i++) {
    // let's just assume that there is a koncorde filter in the current context
    if (this.koncorde.test('pipes', '<event name>', arr[i]).length > 0) {
      filtered.push(arr[i]);
      indices.push(i);
    }
  }

  if (filtered.length === 0) {
    next(null, arr);
    return;
  }

  cb(filtered, (err, res) => {
    if (err)  {
      next(err);
    }
    else {
      // we should test that "res" is an array of the same length than filtered
      for (let i = 0; i < indices.length; i++) {
        arr[i] = res.shift();
      }
    }
  }
}
```